In [2]:
import acids_and_bases as ab
from compound import Compound

# Describe the relative acidity or basicity of a solution

In [3]:
# --- Given values ---
oh_conc = 1e-1  # M

# --- Analysis ---
solution_description = ab.describe_solution_from_oh(oh_conc)

print(solution_description)

The pOH is 1.0 and the pH is 13.0. The solution is highly basic.


# Calculate pH based on ion concentration

In [4]:
# --- Given values ---
h_ion_concentration = 7.9e-6  # M

# --- Calculation ---
pH_value = ab.calculate_pH(h_ion_concentration)

print(f"The pH of the solution is {pH_value:.2f}")


The pH of the solution is 5.10


# Salt is acidic

In [5]:
# --- Salts from the question ---
# Note: The formulas in the image had typos. Using correct formulas.
salts_to_test = {
    "Potassium Formate": "KCHO2", # Corrected from KCO₂H
    "Potassium Bromide": "KBr",
    "Ammonium Bromide": "NH4Br"
}

# --- Analysis ---
print("--- Salt Acidity Analysis ---")
for name, formula in salts_to_test.items():
    acidity = ab.predict_salt_acidity(formula)
    print(f"{name} ({formula}): {acidity}")


--- Salt Acidity Analysis ---
Potassium Formate (KCHO2): Basic
Potassium Bromide (KBr): Neutral
Ammonium Bromide (NH4Br): Acidic


# Rank oxyacids

In [6]:
# --- Acids to compare ---
oxyacids = ["HBrO", "HIO", "HClO"]

# --- Get the ranking ---
ranking = ab.rank_oxyacids_by_strength(oxyacids)

print("The order of oxyacids from strongest to weakest is:")
print(ranking)


The order of oxyacids from strongest to weakest is:
HClO > HBrO > HIO


In [7]:
# --- Acids to compare ---
oxyacids = ["HC2ClH2O2", "HC2BrH2O2", "HC2FH2O2"]

# --- Get the ranking ---
ranking = ab.rank_oxyacids_by_strength(oxyacids)

print("The order of oxyacids from strongest to weakest is:")
print(ranking)


Could not process HC2ClH2O2: Electronegativity for 'C2Cl22' not found.
Could not process HC2BrH2O2: Electronegativity for 'C2Br22' not found.
Could not process HC2FH2O2: Electronegativity for 'C2F22' not found.
The order of oxyacids from strongest to weakest is:



# hydroxide ion concentration based on concentration of h+

In [8]:
# --- Given values ---
h_conc = 1e-2  # M

# --- Calculation ---
oh_conc = ab.calculate_hydroxide_from_hydronium(h_conc)

print(f"The hydroxide ion concentration is {oh_conc:.0e} M")


The hydroxide ion concentration is 1e-12 M


# hydroxide ion concentration based on pOH

In [9]:
# --- Given values ---
pOH_value = 5.60

# --- Calculation ---
hydroxide_concentration = ab.calculate_hydroxide_from_pOH(pOH_value)

print(f"The hydroxide ion concentration is {hydroxide_concentration:.1e} M")

The hydroxide ion concentration is 2.5e-06 M


# Weak Acid Equilibrium (ICE Table)

## pH

In [10]:
# --- Given values ---
initial_concentration = 0.0480 # M
Ka_carbonic_acid = 4.30e-7

# --- Calculation ---
pH = ab.calculate_weak_acid_ph(initial_concentration, Ka_carbonic_acid)

print(f"The pH of the solution is {pH:.1f}")

The pH of the solution is 3.8


## pOH

In [11]:
# --- Given values ---
initial_concentration = 0.31 # M
Kb_ammonia = 1.8e-5

# --- Calculation ---
pOH = ab.calculate_weak_base_pOH(initial_concentration, Kb_ammonia)

print(f"The pOH of the solution is {pOH:.1f}")


The pOH of the solution is 2.6


## Concentration

In [12]:
# --- Given values ---
initial_conc_h2s = 0.0250 # M
Ka_h2s = 1.30e-7

# --- Calculation ---
eq_concentrations = ab.calculate_weak_acid_dissociation(initial_conc_h2s, Ka_h2s)

# The concentration of HS⁻ is the value for the conjugate base, A⁻
hs_concentration = eq_concentrations["A-"]

print(f"The equilibrium concentration of HS⁻ is {hs_concentration:.2e} M")

The equilibrium concentration of HS⁻ is 5.70e-05 M


# Convert K_b to K_a

In [13]:
# --- Given values ---
# This is the Kb for the lactate ion (conjugate base of lactic acid)
kb_lactate = 7.1e-11

# --- Calculation ---
# We calculate the Ka of its conjugate acid (lactic acid)
ka_lactic_acid = ab.convert_ka_kb(kb_lactate)

print(f"The Ka for lactic acid is {ka_lactic_acid:.1e}")

The Ka for lactic acid is 1.4e-04


# Convert K_a to K_b

In [14]:
# --- Given values ---
ka_formic_acid = 1.8e-4

# --- Calculation ---
# We calculate the Kb of its conjugate base (formate)
kb_formate = ab.convert_ka_kb(ka_formic_acid)

print(f"The Kb for the formate ion is {kb_formate:.1e}")

The Kb for the formate ion is 5.6e-11


In [15]:
# --- Given values ---
ka_formic_acid = 6.2e-10

# --- Calculation ---
# We calculate the Kb of its conjugate base (formate)
kb_formate = ab.convert_ka_kb(ka_formic_acid)

print(f"The Kb for the formate ion is {kb_formate:.1e}")

The Kb for the formate ion is 1.6e-05


In [16]:
# --- Given values ---
ka_formic_acid = 7.1e-10

# --- Calculation ---
# We calculate the Kb of its conjugate base (formate)
kb_formate = ab.convert_ka_kb(ka_formic_acid)

print(f"The Kb for the formate ion is {kb_formate:.1e}")

The Kb for the formate ion is 1.4e-05


# Molarity of a solution titration

In [17]:
# --- Given values ---
acid = "HNO3"
base = "Ca(OH)2"

conc_base = 0.175  # M
vol_base = 25.0    # mL
vol_acid = 25.0    # mL

# --- Calculation ---
conc_acid = ab.solve_titration(
    acid_formula=acid, Va_mL=vol_acid,
    base_formula=base, Vb_mL=vol_base,
    known_conc=conc_base, known_is_base=True
)

print(f"The molarity of the {acid} solution is {conc_acid:.3f} M")


The molarity of the HNO3 solution is 0.350 M


In [18]:
# --- Given values ---
acid = "H2SO4"
base = "LiOH"

conc_acid = 3.2      # M
vol_acid = 13.1      # mL
vol_base = 35        # mL

# --- Calculation ---
conc_base = ab.solve_titration(
    acid_formula=acid, Va_mL=vol_acid,
    base_formula=base, Vb_mL=vol_base,
    known_conc=conc_acid, known_is_base=False
)

print(f"The concentration of the {base} solution is {conc_base:.1f} M")

The concentration of the LiOH solution is 2.4 M


In [19]:
# --- Given values ---
acid = "HCl"
base = "KOH"

conc_base = 1.15     # M
vol_base = 45.8      # mL
vol_acid = 20.0      # mL

# --- Calculation ---
conc_acid = ab.solve_titration(
    acid_formula=acid, Va_mL=vol_acid,
    base_formula=base, Vb_mL=vol_base,
    known_conc=conc_base, known_is_base=True
)

print(f"The concentration of the {acid} solution is {conc_acid:.2f} M")


The concentration of the HCl solution is 2.63 M


# [H3O+] of a solution

In [20]:
# --- Given values ---
Ka_HF = 6.3e-4
conc_HF = 0.500
conc_NaF = 0.750

# --- Calculation ---
h3o_concentration = ab.calculate_buffer_h3o_conc(Ka_HF, conc_HF, conc_NaF)

print(f"The [H₃O⁺] of the solution is {h3o_concentration:.1e} M")


The [H₃O⁺] of the solution is 4.2e-04 M


In [21]:
# --- Acids to compare ---
acids = ["HC2F3O2", "HC2FH2O2", "HC2F2HO2"]

# --- Get the ranking ---
ranking = ab.rank_acids_by_inductive_effect(acids)

print("The order of increasing acid strength is:")
print(ranking)

The order of increasing acid strength is:
HC2FH2O2 < HC2F2HO2 < HC2F3O2


# Moles based on volume and moles

In [22]:
# --- Given values ---
molarity_hcl = 0.200  # M
volume_hcl_mL = 15.0  # mL

# The mole ratio from the balanced equation (1 mol LiOH / 1 mol HCl) is 1
mole_ratio = 1

# --- Calculation ---
# 1. Calculate moles of HCl used
moles_hcl = molarity_hcl * (volume_hcl_mL / 1000)

# 2. Use the mole ratio to find moles of LiOH
moles_lih = moles_hcl * mole_ratio

print(f"The moles of LiOH in the sample are {moles_lih:.5f} mol")

The moles of LiOH in the sample are 0.00300 mol


# Calculate pH of a buffer

In [23]:
# --- Given values ---
pKa_carbonic_acid = 6.37
conc_acid = 0.20   # M of H₂CO₃
conc_base = 0.10   # M of NaHCO₃

# --- Calculation ---
pH = ab.calculate_buffer_ph_from_pka(pKa_carbonic_acid, conc_acid, conc_base)

print(f"The pH of the buffer is {pH:.2f}")

The pH of the buffer is 6.07


In [24]:

# --- Given values ---
# Define compounds to get their molar mass
acetic_acid = Compound("HC2H3O2") # Formula for acetic acid
sodium_acetate = Compound("NaC2H3O2") # Formula for sodium acetate

# Set the mass for each to calculate moles
acetic_acid.set_mass(15.01)
sodium_acetate.set_mass(20.51)

# Set the volume to calculate concentration
solution_volume = 1.00 # L
acetic_acid.set_volume(solution_volume)
sodium_acetate.set_volume(solution_volume)

# Ka value
Ka_acetic_acid = 1.80e-5

# --- Calculation ---
pH = ab.calculate_buffer_pH(
    Ka=Ka_acetic_acid,
    conc_acid=acetic_acid.molarity_M,
    conc_base=sodium_acetate.molarity_M
)

print(f"The pH of the buffer solution is {pH:.2f}")

The pH of the buffer solution is 4.74


# Calculate pka from buffer, ph and moles

In [25]:
# --- Given values ---
pH_buffer = 4.41
conc_acid = 0.32  # M
conc_base = 0.14  # M

# --- Calculation ---
pKa_acid = ab.calculate_pka_from_buffer_ph(pH_buffer, conc_acid, conc_base)

print(f"The pKa of the acid is {pKa_acid:.2f}")

The pKa of the acid is 4.77


## The Principle of Buffer Selection
![image.png](images/buffer-1.png)

A buffer solution is most effective at resisting pH changes when the concentrations of the weak acid and its conjugate base are equal. According to the Henderson-Hasselbalch equation, when these concentrations are equal, the pH of the buffer is equal to the pKₐ of the weak acid.

pH=pKa​ (when [Acid] = [Base])

Therefore, to create an effective buffer at a specific pH, you should choose a conjugate acid-base pair where the acid's pKₐ is as close as possible to the target pH.

### Analysis of the Options

We need to find the buffer system whose pKₐ is closest to the target pH of 6.8.

    Sodium hydrogen carbonate (pKₐ = 10.3): Too far from 6.8.

    Acetic acid (pKₐ = 4.8): Too far from 6.8.

    Hypochlorous acid (pKₐ = 7.5): This is the closest pKₐ to the target pH of 6.8.

    Hydrocyanic acid (pKₐ = 9.3): Too far from 6.8.

By choosing the hypochlorous acid system, a chemist can create an effective buffer at pH 6.8 by adjusting the ratio of the conjugate base to the acid.